File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/TechService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.controller.dto.tech.TechAttendanceResponse
import pt.isel.odin.model.Tech

/**
 * Service for techs.
 */
interface TechService {

    /**
     * Gets a tech by its id.
     *
     * @param id the tech id.
     */
    fun getById(id: Long): Tech

    /**
     * Gets all techs.
     */
    fun getAll(): List<Tech>

    /**
     * Saves a tech.
     *
     * @param techRequest the tech info to save without id.
     */
    fun save(techRequest: Tech, email: String): Tech

    /**
     * Updates a tech.
     *
     * @param techRequest the tech info to update with id.
     */
    fun update(techRequest: Tech): Tech

    /**
     * Deletes a tech by its id.
     *
     * @param id the tech id.
     */
    fun delete(id: Long)

    /**
     * Gets all techs by User email.
     *
     * @param email the User email.
     */
    fun getByUser(email: String): List<Tech>

    /**
     * Gets all techs and students by User email.
     *
     * @param email the User email.
     */
    fun getMyTechsAttendance(email: String): List<TechAttendanceResponse>
}

File: code/frontend/Tyr/src/services/TechService.ts
import {makeApiRequest} from '../axios/apiRequest';
import {Tech, TechRequest} from "../model/Tech";

/**
 * Service to manage techs
 */
export class TechService {

    static basePath = '/techs';

    static async getById(id: number): Promise<Tech> {
        return makeApiRequest('get', `${(TechService.basePath)}/${id}`);
    }

    static async getAll(): Promise<Tech[]> {
        return makeApiRequest('get', TechService.basePath);
    }

    static async save(techRequest: TechRequest): Promise<Tech> {
        return makeApiRequest('post', `${(TechService.basePath)}/save`, techRequest);
    }

    static async update(techRequest: TechRequest): Promise<Tech> {
        return makeApiRequest('put', `${(TechService.basePath)}/update`, techRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(TechService.basePath)}/${id}`);
    }

    static async getTechsByUser(): Promise<GetTechsByUserResponse> {
        return makeApiRequest('get', `${(TechService.basePath)}/user`);
    }

/*    static async getMyTechsAttendance(): Promise<TechsAttendance[]> {
        return makeApiRequest('get', `${(TechService.basePath)}/attendance`);
    }*/
}

interface GetTechsByUserResponse {
    techs: Tech[];
}